interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,392 @@
+	if err != nil {
+		return nil, err
+	}
+	for _, listenerReq := range prompt.listenerReqs {
+		if err := sendReply(listenerReq, allow); err != nil {
+			return nil, err
+		}
+	}
+	delete(userEntry.ByID, id)
+	data := map[string]string{"resolved": "replied"}
+	pdb.notifyPrompt(user, id, data)
+	return prompt, nil
+}
+
+var sendReply = func(listenerReq *listener.Request, reply interface{}) error {
+	return listenerReq.Reply(reply)
+}
+
+// HandleNewRule checks if any existing prompts are satisfied by the given rule
+// contents and, if so, sends back a decision to their listener requests.
+//
+// A prompt is satisfied by the given rule contents if the user, snap,
+// interface, and path of the prompt match those of the rule, and if either the
+// outcome is "allow" and all of the prompt's permissions are matched by those
+// of the rule contents, or if the outcome is "deny" and any of the permissions
+// match.
+//
+// Records a notice for any prompt which was satisfied, or which had some of
+// its permissions satisfied by the rule contents. In the future, only the
+// remaining unsatisfied permissions of a partially-satisfied prompt must be
+// satisfied for the prompt as a whole to be satisfied.
+//
+// Returns the IDs of any prompts which were fully satisfied by the given rule
+// contents.
+func (pdb *PromptDB) HandleNewRule(metadata *prompting.Metadata, constraints *prompting.Constraints, outcome prompting.OutcomeType) ([]string, error) {
+	allow, err := outcome.AsBool()
+	if err != nil {
+		return nil, err
+	}
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[metadata.User]
+	if !ok {
+		return nil, nil
+	}
+	var satisfiedPromptIDs []string
+	for id, prompt := range userEntry.ByID {
+		if !(prompt.Snap == metadata.Snap && prompt.Interface == metadata.Interface) {
+			continue
+		}
+		matched, err := constraints.Match(prompt.Constraints.Path)
+		if err != nil {
+			return nil, err
+		}
+		if !matched {
+			continue
+		}
+		modified := prompt.Constraints.subtractPermissions(constraints.Permissions)
+		if !modified {
+			continue
+		}
+		if len(prompt.Constraints.Permissions) > 0 && allow == true {
+			pdb.notifyPrompt(metadata.User, id, nil)
+			continue
+		}
+		// All permissions of prompt satisfied, or any permission denied
+		for _, listenerReq := range prompt.listenerReqs {
+			sendReply(listenerReq, allow)
+		}
+		delete(userEntry.ByID, id)
+		satisfiedPromptIDs = append(satisfiedPromptIDs, id)
+		data := map[string]string{"resolved": "satisfied"}
+		pdb.notifyPrompt(metadata.User, id, data)
+	}
+	return satisfiedPromptIDs, nil
+}
+
+// Close removes all outstanding prompts and records a notice for each one.
+//
+// This should be called when snapd is shutting down, to notify prompt clients
+// that the given prompts are no longer awaiting a reply.
+func (pdb *PromptDB) Close() {
+	data := map[string]string{"resolved": "cancelled"}
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
```

Comment:
Let's Unmap the memory we've got as this technically holds an open file description and is not a pure memory resource.